R/xgb_binning.R

```diff
@@ -0,0 +1,250 @@
+#'  by their original name.
+#'
+#' @examples
+#' data(credit_data)
+#' library(rsample)
+#'
+#' split <- initial_split(credit_data, strata = "Status")
+#'
+#' credit_data_tr <- training(split)
+#' credit_data_te <- testing(split)
+#'
+#' xgb_rec <- recipe(Status ~ ., data = credit_data_tr) %>%
+#'   step_medianimpute(all_numeric()) %>%
+#'   step_xgb_binning(all_numeric(), outcome = "Status")
+#'
+#' xgb_rec <- prep(xgb_rec, training = credit_data_tr, retain = TRUE)
+#'
+#' xgb_test_bins <- bake(xgb_rec, credit_data_te)
+#' @seealso [recipe()] [prep.recipe()] [bake.recipe()]
+#'
+
+step_xgb_binning <-
+  function(recipe,
+           ...,
+           role = NA,
+           trained = FALSE,
+           outcome = NULL,
+           rules = NULL,
+           prefix = "xgb_bin",
+           skip = FALSE,
+           id = rand_id("xgb_binning")) {
+    if (is.null(outcome))
+      stop("`outcome` should select at least one column.", call. = FALSE)
+
+    # recipes_pkg_check("xgb_binninb") - TO DO CHECKS
+
+    add_step(
+      recipe,
+      step_xgb_binning_new(
+        terms = ellipse_check(...),
+        role = role,
+        trained = trained,
+        outcome = outcome,
+        rules = rules,
+        prefix = prefix,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_xgb_binning_new <-
+  function(terms, role, trained, outcome, rules,
+           prefix, skip, id) {
+    step(
+      subclass = "xgb_binning",
+      terms = terms,
+      role = role,
+      trained = trained,
+      outcome = outcome,
+      rules = rules,
+      prefix = prefix,
+      skip = skip,
+      id = id
+    )
+  }
+
+run_xgboost <- function(.train, .test, .max_bin){
+  xgboost::xgb.train(
+    params = list(
+      eta = 0.3,
+      max_bin = .max_bin,
+      max_depth = 1,
+      min_child_weight = 5
+    ),
+    nrounds = 100,
+    data = .train,
+    watchlist = list(
+      train = .train,
+      test  = .test
+    ),
+    tree_method = "hist",
+    early_stopping_rounds = 10,
+    objective = "binary:logistic",
+    eval_metric = "auc",
+    verbose = 0
+  )
+}
+
+xgb_binning <- function(df, target, variable){
+
+  levels <- sort(unique(df[[target]]))
+
+  if(length(levels) != 2){
+    stop("Target variable does not have exactly two levels.", call. = FALSE)
+  }
+
+  if(any(is.na((levels)))){
+    stop("Target variable contains missing values.", call. = FALSE)
+  }
+
+  split <- rsample::initial_split(df, prop = 0.8, strata = target)
+  train <- rsample::training(split)
+  test  <- rsample::testing(split)
+
+  xgb_train <- xgboost::xgb.DMatrix(
+    data = as.matrix(train[, variable]),
+    label = ifelse(train[[target]] == levels[[1]], 0, 1)
+  )
+
+  xgb_test <- xgboost::xgb.DMatrix(
+    data = as.matrix(test[, variable]),
+    label = ifelse(test[[target]] == levels[[1]], 0, 1)
+  )
+
+  xgb_runs <- purrr::map(c(10, 20, 50), function(x) run_xgboost(xgb_train, xgb_test, x))
```

Comment:
Instead of hard-coding these, I would set it up for one value and we can tune it as needed. This will simplify the code a lot too. Please keep the argument names consistent with `parsnip:::boost_tree()`

Reply:
After the changes I made this part looks like this now:

```
xgb_mdl <- withr:::with_seed(sample.int(10^6, 1), run_xgboost(xgb_train, xgb_test, num_breaks, objective))
```

What would you change here specifically in the context of ```parsnip:::boost_tree()```?

Reply:
I'm not sure; I would treat `num_breaks` as if it were any other parameter. Maybe I need to kick this around a bit to make a good assessment.

Reply:
Implemented in line with the comment about parameters I made above.